File: largest_rectangle/largest_rectangle.ts
export {};

function largestRectangle(heights: number[]): number {
  const stack: number[] = [];
  let maxArea: number = 0;
  const n = heights.length;

  for (let i = 0; i < n; i++) {
    while (stack.length !== 0 && heights[i] < heights[stack[stack.length - 1]]) {
      const h = heights[stack.pop()!];
      const width = stack.length === 0 ? i : i - stack[stack.length - 1] - 1;
      maxArea = Math.max(maxArea, width * h);
    }
    stack.push(i);
  }

  while (stack.length !== 0) {
    const h = heights[stack.pop()!];
    const width = stack.length === 0 ? n : n - stack[stack.length - 1] - 1;
    maxArea = Math.max(maxArea, h * width);
  }

  return maxArea;
}

const heights = [2, 1, 5, 6, 2, 3];
console.log("The max area of the rectangle is: ", largestRectangle(heights));

File: largest_rectangle/largest_rectangle.py
def largest_rectangle(heights):
    max_area = 0
    stack = []
    n = len(heights)
    
    for i in range(n):
        while stack and heights[i] < heights[stack[-1]]:
            h = heights[stack.pop()]
            width = i if not stack else i - stack[-1] - 1
            max_area = max(max_area, h * width)
        stack.append(i)
        
    while stack:
        h = heights[stack.pop()]
        width = n if not stack else n - stack[-1] - 1
        max_area = max(max_area, h * width)
        
    return max_area

heights = [2, 1, 5, 6, 2, 3]
print("The max area of the rectangle is: ", largest_rectangle(heights))